test/functional/inputs/http/http_input_test.go

```diff
@@ -0,0 +1,137 @@
+	. "github.com/onsi/gomega"
+	logging "github.com/openshift/cluster-logging-operator/apis/logging/v1"
+	"github.com/openshift/cluster-logging-operator/internal/constants"
+	"github.com/openshift/cluster-logging-operator/internal/runtime"
+	"github.com/openshift/cluster-logging-operator/internal/utils"
+	"github.com/openshift/cluster-logging-operator/test/framework/functional"
+	testfw "github.com/openshift/cluster-logging-operator/test/functional"
+	"github.com/openshift/cluster-logging-operator/test/helpers/types"
+	. "github.com/openshift/cluster-logging-operator/test/matchers"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	servicePort = 8080
+)
+
+var _ = Describe("[Functional][Inputs][Http] Functional tests", func() {
+
+	tsF := func(tsStr string) time.Time { ts, _ := time.Parse(time.RFC3339Nano, tsStr); return ts }
+
+	var (
+		framework   *functional.CollectorFunctionalFramework
+		auditRecord = types.K8sAuditLog{
+			AuditLogCommon: types.AuditLogCommon{
+				Kind:                     `Event`,
+				APIVersion:               `audit.k8s.io/v1`,
+				Hostname:                 `*`,
+				LogType:                  `audit`,
+				Level:                    `default`,
+				AuditID:                  `16dcc977-a39e-467b-af44-921416b7800b`,
+				Stage:                    `RequestReceived`,
+				RequestURI:               `/apis/rbac.authorization.k8s.io/v1/clusterrolebindings?limit=500\u0026resourceVersion=0`,
+				Timestamp:                tsF(`2022-08-17T20:27:20.570375Z`),
+				RequestReceivedTimestamp: tsF(`2023-08-24T21:37:38.842649Z`),
+				ViaqMsgID:                "*",
+				PipelineMetadata:         functional.TemplateForAnyPipelineMetadata,
+				OpenshiftLabels: types.OpenshiftMeta{
+					ClusterID: "*",
+					Sequence:  types.NewOptionalInt(""),
+				},
+			},
+			K8SAuditLevel: "default",
+		}
+	)
+
+	BeforeEach(func() {
+		Expect(testfw.LogCollectionType).To(Equal(logging.LogCollectionTypeVector))
+		framework = functional.NewCollectorFunctionalFrameworkUsingCollector(logging.LogCollectionTypeVector)
+		functional.NewClusterLogForwarderBuilder(framework.Forwarder).
+			FromInputWithVisitor("http-source",
+				func(spec *logging.InputSpec) {
+					spec.Source = &logging.SourceSpec{
+						HTTP: &logging.HTTPSource{
+							Name: `http-source-spec-name`,
+							Port: v1.ServicePort{
+								Port: servicePort,
+							},
+							Format:  logging.FormatK8S,
+							LogType: `audit`,
+						},
+					}
+				}).ToHttpOutput()
+	})
+
+	AfterEach(func() {
+		framework.Cleanup()
+	})
+
+	Context("When sending an audit log record to an HTTP input", func() {
+		It("should be able to round trip it unharmed", func() {
+			Expect(framework.DeployWithVisitor(
+				func(b *runtime.PodBuilder) error {
+					return framework.AddVectorHttpOutput(b, framework.Forwarder.Spec.Outputs[0])
+				}),
+			).To(BeNil())
+
+			_, err := framework.RunCommand(constants.CollectorName, "curl", "-ksv", fmt.Sprintf("http://localhost:%d", servicePort), "-d",
```

Comment:
looks like a method to be added to write.go and something with a convenience method to write to the named input

Reply:
On my audit branch I added a LogWriter and WriteLog to write.go, they  use a single "cat" command to pipe data directly to a log file. It is vastly faster than execing a a separate "echo" per line, and it keeps all the log generation in Go rather than quoted shell script parmeters, so avoids all the shell quoting madness:

https://github.com/alanconway/cluster-logging-operator/blob/api-audit-filter/test/framework/functional/write.go#L179

I'd suggest something similar here - i.e. rather than a curl command, set up a port-forward to the HTTP address and then use the Go http client to post directly.

Reply:
@alanconway Done ^^^. Changes are in
test/client/client.go
test/framework/functional/read.go
test/framework/functional/write.go
test/functional/inputs/http/http_input_test.go